### ISSUE TITLE
if have parallel lines

### ISSUE THREAD

**Issue by wandh** [snatik/polygon-contains-point#3]
Polygon p= Polygon.Builder().addVertex(new Point(0, 0)).addVertex(new Point(0,1)).addVertex(new Point(1,2)).addVertex(new Point(1,99)).addVertex(new Point(100,0)).build();
        System.out.println(p.contains(new Point(3f,4f)));//false  :error
        System.out.println(p.contains(new Point(3f,4.1f)));//true
        System.out.println(p.contains(new Point(3f,3.9f)));//true

**Comment by tatarize** [comment]
```
			// check if both vectors are parallel. If they are parallel then no intersection point will exist
			if (ray.getA() - side.getA() == 0)
			{
				return false;
			}
```

Yeah, in your case those two lines have the same slope but the point being checked is notably well within the polygon. It's trying to see if the two lines have the same slope and they do. It assumes since the two lines are parallel they don't meet, but it doesn't account for the ray escaping the polygon through a parallel line. Just that the line on the edge of the polygon if it continued on would be on that same line, needed for the ray to reach the end of the bounding box.  And since isInside for the line calls adds one for each point when it's equal, you end up with an escape score of 2. 

```
if ((point.x >= minX && point.x <= maxX) && (point.y >= minY && point.y <= maxY))
		{
			return true;
		}
		return false;
```

Properly this also means that any rays that attempts escape by exactly hitting a vertex is also said to be outside the polygon. In most accountings of the algorithm the solution here is to pick endpoint to be inside the line and another endpoint to be outside the line. 

The correct function for isInside() should read: 

        
```
        public boolean isInside(Point point)
        {
            double distanceSP = _start.distanceSq(point);
            if (distanceSP == 0) return false;
            return ((distanceSP + _end.distanceSq(point)) == _start.distanceSq(_end));
        }
```

Assuming Point has that function:

`
```
public double distanceSq(Point p) {
            double dx = p.x - x;
            double dy = p.y - y;
            dx *= dx;
            dy *= dy;
            return dx + dy;
        }`
```

**Comment by tatarize** [comment]
The vertex bug is because if you draw a ray through a vertex it counts as two intersections one for each line that shares that vertex. My function firstly actually checks if that point is on the line rather than within a bounding box that captures that line (which for your uses isn't wrong) but checks if the end point is coincident with the point being checked and counts that as false. You can equally fix original code by just taking the first line here. As long as you say that a point coincident with either a start point or end point, but not both counts as zero we maintain odd parity for insideness.

```
	public boolean isInside(Point point)
    {
        if ((point.x == _end.x) && (point.y == _end.y)) return false;
        if (_end.x == _start.x) {
            float ty = (point.y - _start.y) / (_end.y - _start.y);
            return (ty >= 0) && (ty <= 1);
        }
        if (_end.y == _start.y) {
            float tx = (point.x - _start.x) / (_end.x - _start.x);
            return (tx >= 0) && (tx <= 1);
        }
        float tx = (point.x - _start.x) / (_end.x - _start.x);
        float ty = (point.y - _start.y) / (_end.y - _start.y);
        return (tx == ty) && (tx >= 0) && (tx <= 1);
	}
```

**Comment by tatarize** [comment]
It isn't that you escaped the ray through a parallel line, it's that you escaped that ray through a vertex.  It counts the end of the first line as 1 and the start of the line intercepting the parallel line as 1. And 0 points for the parallel line. But, you'll hit the same issue if you discount parallel lines and escape through any old vertex.

**Comment by sromku** [comment]
@wandh The issue was with ray calculation points. It's fixed. 

I added your example as test case - [testParallel()](https://github.com/snatik/polygon-contains-point/blob/master/src/test/java/Tests.java#L145)

You can use latest release **1.0.2** with the fixes - [polygon-1.0.2.jar](https://github.com/sromku/polygon-contains-point/releases/download/1.0.2/polygon-1.0.2.jar)

Much thanks!
